implementations/rust/ockam/ockam/src/entity/profile/profile_helpers.rs

```diff
@@ -0,0 +1,144 @@
+use crate::entity::{Profile, ProfileChangeEvent};
+use crate::OckamError;
+use ockam_vault_core::secret::Secret;
+
+impl Profile {
+    pub(crate) fn get_last_event_id(&self) -> Result<EventIdentifier, ockam_core::Error> {
+        if let Some(e) = self.change_history.as_ref().last() {
+            Ok(e.identifier().clone())
+        } else {
+            Err(OckamError::InvalidInternalState.into())
+        }
+    }
+
+    pub(crate) fn find_key_change_in_event<'a>(
+        event: &'a ProfileChangeEvent,
+        label: &str,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+    ) -> Option<&'a ProfileChange> {
+        event
+            .changes()
+            .as_ref()
+            .iter()
+            .rev()
+            .find(|c| match c.change_type() {
+                CreateKey(change) => {
+                    change.key_type() == key_type
+                        && change.key_purpose() == key_purpose
+                        && change.label() == label
+                }
+                RotateKey(change) => {
+                    change.key_type() == key_type
+                        && change.key_purpose() == key_purpose
+                        && change.label() == label
+                } // RevokeKey(event) => {
+                  //     event.key_type() == key_type && event.key_purpose() == key_purpose && event.label() == label
+                  // }
+            })
+    }
+
+    pub(crate) fn find_last_key_event(
+        &self,
+        label: &str,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+    ) -> Result<&ProfileChangeEvent, ockam_core::Error> {
+        self.change_history
+            .as_ref()
+            .iter()
+            .rev()
+            .find(|e| Self::find_key_change_in_event(e, label, key_type, key_purpose).is_some())
+            .ok_or(OckamError::InvalidInternalState.into())
+    }
+
+    pub(crate) fn find_key_event_before(
+        &self,
+        event_id: &EventIdentifier,
+        label: &str,
+        key_type: ProfileKeyType,
+        key_purpose: ProfileKeyPurpose,
+    ) -> Result<&ProfileChangeEvent, ockam_core::Error> {
+        let before_index = self
+            .change_history
+            .as_ref()
+            .iter()
+            .position(|e| e.identifier() == event_id)
+            .unwrap_or(self.change_history.as_ref().len());
+        self.change_history.as_ref()[..before_index]
+            .iter()
+            .rev()
+            .find(|e| Self::find_key_change_in_event(e, label, key_type, key_purpose).is_some())
+            .ok_or(OckamError::InvalidInternalState.into())
+    }
+
+    pub(crate) fn get_change_public_key(
+        change: &ProfileChange,
+    ) -> Result<&[u8], ockam_core::Error> {
+        match change.change_type() {
+            CreateKey(change) => Ok(change.public_key()),
+            RotateKey(change) => Ok(change.public_key()),
+        }
+    }
+
+    pub(crate) fn get_private_key_from_event(
```

Comment:
```suggestion
    pub(crate) fn get_secret_key_from_event(
```

Reply:
@mikelodder7 do we have a place with decisions about such naming? If not, we should create one